File: macos/Runner/Capture/Zoom/ZoomFollowSmoother.swift
import CoreGraphics
import Foundation

enum ZoomFollowSmoother {
  static let minStrength: CGFloat = 0.05
  static let maxStrength: CGFloat = 0.50
  static let defaultStrength: CGFloat = 0.15

  static let minDtSeconds: Double = 1.0 / 240.0
  static let maxDtSeconds: Double = 1.0 / 15.0
  static let defaultReferenceFPS: Double = 60.0

  static func clampedFollowStrength(_ value: CGFloat) -> CGFloat {
    let source = value.isFinite ? value : defaultStrength
    return min(max(source, minStrength), maxStrength)
  }

  static func alpha(
    baseStrength: CGFloat,
    dt: Double,
    referenceFPS: Double = defaultReferenceFPS
  ) -> CGFloat {
    let strength = Double(clampedFollowStrength(baseStrength))
    let clampedDt = clampedDtSeconds(dt)
    let safeReferenceFPS =
      (referenceFPS.isFinite && referenceFPS > 0) ? referenceFPS : defaultReferenceFPS
    let normalizedFrames = clampedDt * safeReferenceFPS
    let alphaValue = 1.0 - pow(1.0 - strength, normalizedFrames)
    return CGFloat(min(max(alphaValue, 0.0), 1.0))
  }

  static func lerp(current: CGFloat, target: CGFloat, alpha: CGFloat) -> CGFloat {
    let clampedAlpha = min(max(alpha.isFinite ? alpha : 0, 0), 1)
    return current + (target - current) * clampedAlpha
  }

  static func clampedDtSeconds(_ value: Double) -> Double {
    let source = value.isFinite ? value : (1.0 / defaultReferenceFPS)
    return min(max(source, minDtSeconds), maxDtSeconds)
  }
}

enum ZoomFollowParityDebug {
  private static let envValue: String =
    ProcessInfo.processInfo.environment["CLINGFY_ZOOM_PARITY_DEBUG"] ?? ""

  static let enabled: Bool = {
    let normalized = envValue.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    return normalized == "1" || normalized == "true" || normalized == "yes"
  }()

  static func shouldLogPreview(tick: Int) -> Bool {
    enabled && tick % 30 == 0
  }

  static func shouldLogExport(frameIndex: Int) -> Bool {
    enabled && frameIndex % 30 == 0
  }

  static func logSample(
    source: String,
    time: Double,
    zoom: CGFloat,
    centerX: CGFloat,
    centerY: CGFloat,
    targetZoom: CGFloat,
    targetCenterX: CGFloat,
    targetCenterY: CGFloat
  ) {
    guard enabled else { return }
    NativeLogger.d(
      "ZoomParity",
      "Zoom follow sample",
      context: [
        "source": source,
        "time": time,
        "zoom": zoom,
        "centerX": centerX,
        "centerY": centerY,
        "targetZoom": targetZoom,
        "targetCenterX": targetCenterX,
        "targetCenterY": targetCenterY,
      ]
    )
  }
}

enum CameraPlacementDebug {
  private static let envValue: String =
    ProcessInfo.processInfo.environment["CLINGFY_CAMERA_PLACEMENT_DEBUG"] ?? ""

  static let enabled: Bool = {
    let normalized = envValue.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    return normalized == "1" || normalized == "true" || normalized == "yes"
  }()

  static func shouldLogPreview(tick: Int) -> Bool {
    enabled && (tick <= 1 || tick % 30 == 0)
  }

  static func shouldLogExport(
    sampleIndex: Int,
    sampleCount: Int,
    isFirstActiveZoomSample: Bool
  ) -> Bool {
    enabled
      && (
        sampleIndex == 0
          || sampleIndex == max(sampleCount - 1, 0)
          || sampleIndex % 30 == 0
          || isFirstActiveZoomSample
      )
  }

  static func rectContext(prefix: String, rect: CGRect?) -> [String: Any] {
    guard let rect else {
      return ["\(prefix)Present": false]
    }
    return [
      "\(prefix)Present": true,
      "\(prefix)X": rect.origin.x,
      "\(prefix)Y": rect.origin.y,
      "\(prefix)Width": rect.size.width,
      "\(prefix)Height": rect.size.height,
    ]
  }

  static func sizeContext(prefix: String, size: CGSize) -> [String: Any] {
    [
      "\(prefix)Width": size.width,
      "\(prefix)Height": size.height,
    ]
  }

  static func pointContext(prefix: String, point: CGPoint?) -> [String: Any] {
    guard let point else {
      return ["\(prefix)Present": false]
    }
    return [
      "\(prefix)Present": true,
      "\(prefix)X": point.x,
      "\(prefix)Y": point.y,
    ]
  }

  static func transformContext(prefix: String, transform: CGAffineTransform) -> [String: Any] {
    [
      "\(prefix)A": transform.a,
      "\(prefix)B": transform.b,
      "\(prefix)C": transform.c,
      "\(prefix)D": transform.d,
      "\(prefix)Tx": transform.tx,
      "\(prefix)Ty": transform.ty,
    ]
  }
}

File: windows/runner/Capture/Zoom/zoom_follow_smoother.cpp
#include "Capture/Zoom/zoom_follow_smoother.h"

#include <cmath>

#include "preview/zoom_easing_constants.h"

namespace clingfy::capture {

namespace {
bool IsFinite(double v) { return std::isfinite(v); }
}  // namespace

double ClampFollowStrength(double value) {
  const double source = IsFinite(value)
                            ? value
                            : clingfy::preview::kZoomFollowStrengthDefault;
  if (source < clingfy::preview::kZoomFollowStrengthMin) {
    return clingfy::preview::kZoomFollowStrengthMin;
  }
  if (source > clingfy::preview::kZoomFollowStrengthMax) {
    return clingfy::preview::kZoomFollowStrengthMax;
  }
  return source;
}

double ClampFollowDtSeconds(double value) {
  const double source =
      IsFinite(value) ? value : (1.0 / clingfy::preview::kZoomFollowReferenceFPS);
  if (source < clingfy::preview::kZoomFollowMinDtSeconds) {
    return clingfy::preview::kZoomFollowMinDtSeconds;
  }
  if (source > clingfy::preview::kZoomFollowMaxDtSeconds) {
    return clingfy::preview::kZoomFollowMaxDtSeconds;
  }
  return source;
}

double ZoomFollowAlpha(double base_strength, double dt_seconds,
                       double reference_fps) {
  const double strength = ClampFollowStrength(base_strength);
  const double clamped_dt = ClampFollowDtSeconds(dt_seconds);
  const double safe_fps = (IsFinite(reference_fps) && reference_fps > 0.0)
                              ? reference_fps
                              : clingfy::preview::kZoomFollowReferenceFPS;
  const double normalized_frames = clamped_dt * safe_fps;
  const double alpha = 1.0 - std::pow(1.0 - strength, normalized_frames);
  if (alpha < 0.0) return 0.0;
  if (alpha > 1.0) return 1.0;
  return alpha;
}

double ZoomFollowLerp(double current, double target, double alpha) {
  double a = IsFinite(alpha) ? alpha : 0.0;
  if (a < 0.0) a = 0.0;
  if (a > 1.0) a = 1.0;
  return current + (target - current) * a;
}

}  // namespace clingfy::capture
